File: PixelSpriteEditor/src/PencilTool.java
/**
 * Freehand drawing. A drag is treated as one continuous stroke (one undo
 * step); gaps between fast pointer movements are filled with a line so no
 * pixels are skipped.
 */
public class PencilTool implements Tool {
    private int lastX, lastY;
    private boolean hasLast;

    @Override
    public void onPress(EditorState state, int x, int y) {
        state.beginStroke();
        state.paintPixel(x, y, state.getCurrentColorArgb());
        lastX = x;
        lastY = y;
        hasLast = true;
    }

    @Override
    public void onDrag(EditorState state, int x, int y) {
        if (!hasLast) {
            lastX = x;
            lastY = y;
            hasLast = true;
        }
        int argb = state.getCurrentColorArgb();
        LineUtil.forEachLinePixel(lastX, lastY, x, y, (px, py) -> state.paintPixel(px, py, argb));
        lastX = x;
        lastY = y;
    }

    @Override
    public void onRelease(EditorState state, int x, int y) {
        state.endStroke();
        hasLast = false;
    }
}

File: legacy-csharp-avalonia/src/Tools/PencilTool.cs
namespace PixelSpriteEditor.Tools;

/// <summary>
/// Freehand drawing. A drag is treated as one continuous stroke (one undo
/// step); gaps between fast pointer movements are filled with a line so no
/// pixels are skipped.
/// </summary>
public sealed class PencilTool : ITool
{
    private int _lastX, _lastY;
    private bool _hasLast;

    public void OnPointerDown(IEditorContext ctx, int x, int y)
    {
        ctx.BeginStroke();
        ctx.PaintPixel(x, y, ctx.CurrentColor);
        _lastX = x;
        _lastY = y;
        _hasLast = true;
    }

    public void OnPointerMove(IEditorContext ctx, int x, int y)
    {
        if (!_hasLast) { _lastX = x; _lastY = y; _hasLast = true; }
        var color = ctx.CurrentColor;
        LineUtil.ForEachLinePixel(_lastX, _lastY, x, y, (px, py) => ctx.PaintPixel(px, py, color));
        _lastX = x;
        _lastY = y;
    }

    public void OnPointerUp(IEditorContext ctx, int x, int y)
    {
        ctx.EndStroke();
        _hasLast = false;
    }
}
